internal/lifecycle/state.go

```diff
@@ -0,0 +1,55 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lifecycle // import "go.opentelemetry.io/collector/internal/lifecycle"
+
+import "go.uber.org/atomic"
+
+// Lifecycle represents the current
+// running state of the component
+// that is intended to be concurrency safe.
+type State interface {
+	// Start atomically updates the state and returns true if
+	// the previous state was not running.
+	Start() (started bool)
+
+	// Stop atomically updates the state and returns true if
+	// the previous state was not shutdown.
+	Stop() (stopped bool)
+}
```

Comment:
```suggestion
// This will implement component.Component.
type Component struct {
	current *atomic.Int64
	component.StartFunc
	component.ShutdownFunc
}

func NewComponent(start component.StartFunc, shutdown component.ShutdownFunc) *Component {
	return &Component{
		current: atomic.NewInt64(shutdown),
		StartFunc: start,
		ShutdownFunc: shutdown
	}
}

func (c *Component) Start(ctx context.Context, host component.Host) error {
	if s.current.Swap(running) != shutdown {
		return nil
	}
	c.StartFunc.Start(ctx, host)
}

func (c *Component) Shutdown(ctx context.Context) error {
	if s.current.Swap(running) != running {
		return nil
	}
	c.ShutdownFunc.Shutdown(ctx)
}
```

Reply:
Then in the exporterhelper can have 

```
type baseExporter struct {
	*lifecyle.Component
}
```